estate/models/estate_property.py

```diff
@@ -0,0 +1,121 @@
+        selection=[
+            ('north', 'North'),
+            ('south', 'South'),
+            ('east', 'East'),
+            ('west', 'West'),
+        ],
+    )
+    active = fields.Boolean(default=True)
+    state = fields.Selection(
+        string='State',
+        required=True,
+        default='new',
+        selection=[
+            ('new', 'New'),
+            ('offer_received', 'Offer Received'),
+            ('offer_accepted', 'Offer Accepted'),
+            ('sold', 'Sold'),
+            ('cancelled', 'Cancelled'),
+        ],
+        )
+    property_type_id = fields.Many2one(
+        'estate.property.type', string='Property Type',
+    )
+    buyer_id = fields.Many2one(
+        'res.partner', string='Buyer', copy=False,
+    )
+    salesperson_id = fields.Many2one(
+        'res.users', string='Salesperson',
+        default=lambda self: self.env.user,
+    )
+    tag_ids = fields.Many2many('estate.property.tag', string='Tags')
+    offer_ids = fields.One2many('estate.property.offer', 'property_id', string='Offers')
+
+    _check_expected_price = models.Constraint(
+        'CHECK(expected_price > 0)',
+        'The expected price of a property should be strictly positive.',
+    )
+    _check_selling_price = models.Constraint(
+        'CHECK(selling_price >= 0)',
+        'The selling price of a property should be positive.',
+    )
+
+    @api.depends('living_area', 'garden_area')
+    def _compute_total_area(self):
+        for record in self:
+            living_area = record.living_area or 0
+            garden_area = record.garden_area or 0
+            record.total_area = living_area + garden_area
+
+    @api.depends('offer_ids')
+    def _compute_best_price(self):
+        for record in self:
+            offers = record.offer_ids or []
+            record.best_price = max(offers.mapped('price')) if len(offers) else 0
+
+    @api.onchange('garden')
+    def _onchange_garden(self):
+        if self.garden:
+            self.garden_area = 10
+            self.garden_orientation = 'north'
+        else:
+            self.garden_area = None
+            self.garden_orientation = None
+
+    @api.constrains('selling_price', 'expected_price')
+    def _check_selling_price(self):
+        for record in self:
+            if (
+                not float_is_zero(record.selling_price, 2)
+                and float_compare(record.selling_price, record.expected_price * 0.90, 2) < 0
+            ):
+                raise ValidationError("Selling price cannot be lower than 90% of expected price")
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_if_new_or_cancelled(self):
+        for record in self:
+            if record.state not in ['new', 'cancelled']:
+                raise UserError(f"Can't delete property in {record.state} state")
+
+    def action_mark_as_sold(self):
+        for record in self:
+            if record.state == 'cancelled':
+                raise UserError('Property is already cancelled')
```

Comment:
```suggestion
                raise UserError('Cannot sell cancelled properties')
```
For errors it is better to be precise about why can't we do the action not just that we can't do it

Reply:
Thank you. That's a good point. Error messages updated.